source/core_mqtt.c

```diff
@@ -737,10 +748,14 @@ static int32_t recvExact( const MQTTContext_t * pContext,
                         ( unsigned long ) bytesRemaining,
                         ( long int ) totalBytesRecvd ) );
         }
+        else
+        {
+            /* No bytes were read from the network. */
+            timeSinceLastRecvMs = calculateElapsedTime( getTimeStampMs(), lastDataRecvTimeMs );
+        }
 
-        elapsedTimeMs = calculateElapsedTime( getTimeStampMs(), entryTimeMs );
-
-        if( ( bytesRemaining > 0U ) && ( elapsedTimeMs >= timeoutMs ) )
+        if( ( bytesRemaining > 0U ) &&
```

Comment:
Should this if(){} be inside the else{} immediately above it so it is only checked if no bytes are received?  That would also prevent timeSinceLastRecvMs being tested while it had its initialisation value of zero (which would only be a problem if MQTT_RECV_POLLING_TIMEOUT_MS was also zero).

Reply:
Nesting the if() within the above immediate else is ideal, but that increases the complexity score of the `recvExact` function beyond the threshold score of 8 (thereby, causing CI check failure). Keeping the complexity score <=8  is the only reason of not having the if-check nested within the else.

Do we want the behavior of retrying the transport recv call once even when the `MQTT_RECV_POLLING_TIMEOUT_MS` value is set to 0? 
The current behavior will not retry if the `MQTT_RECV_POLLING_TIMEOUT_MS` is set to 0 and the first call to transport recv returned 0 bytes.

Reply:
Ref the complexity score - I understand that is the requirement but in this case it is leading to code that is not ideal so perhaps we should reduce complexity in another way?

What do you think to just prohibiting MQTT_RECV_POLLING_TIMEOUT_MS from being 0 with a #error message if it is.  We only go into the recvExact() function when we know there is data to receive - and exiting again without that data will just cause an MQTT reset.  In the case where we are just reading one byte to determine if there is any data we don't want to block - but in this case we already know there is data.

Reply:
>Ref the complexity score - I understand that is the requirement but in this case it is leading to code that is not ideal so perhaps we should reduce complexity in another way?

Have reduced the complexity from log message to relocate the if-check in its ideal place :)

Reply:
Gread - no need to restrict MQTT_RECV_POLLING_TIMEOUT_MS to non zero values.